src/braft/snapshot.cpp

```diff
@@ -777,8 +785,30 @@ void LocalSnapshotCopier::copy() {
         }
         std::vector<std::string> files;
         _remote_snapshot.list_files(&files);
-        for (size_t i = 0; i < files.size() && ok(); ++i) {
-            copy_file(files[i]);
+
+        // Do the copy in chunks. We don't want to submit 1 million concurrent
+        // GetFileRequests to the remote
+        // server. This is configurable via
+        // raft_max_get_file_request_concurrency. Making a copy of this onto the
+        // stack in case it changes while this loop is running.
+        std::span<std::string> files_span{files};
+        std::size_t chunk_size = FLAGS_raft_max_get_file_request_concurrency;
+        for (std::size_t i = 0; i < files_span.size(); i += chunk_size) {
+          {
+            BAIDU_SCOPED_LOCK(_mutex);
+            if (_cancelled) {
+              break;
+            }
+          }
+
+          if (!ok()) {
+            break;
+          }
+
+          auto remaining = files_span.size() - i;
+          auto current_chunk_size = std::min(chunk_size, remaining);
+          auto chunk = files_span.subspan(i, current_chunk_size);
+          copy_files(chunk);
```

Comment:
Copy 64(default value) files first, and then the next 64 files?
It seems not a good concurrency strategy.

Reply:
You're right it is not ideal, but it is still a dramatic improvement and would have prevented a production incident recently.

I can see if there is an easy way to try to send requests as others complete so we don't do it in chunks. I'm worried that will require more fundamental changes since. You can't just call Session::join() because you don't know which request will finish first. You could repeatedly poll Session::status() but that would be a waste of CPU. I think a callback / continuation would have to be added to Session to do this properly.

Reply:
I guess another alternative is to just send all of them concurrently, but enable the snapshot throttle.

Reply:
I got you. If `RemoteFileCopier::Session` has such routine as `on_finish` we can have a better solution.
But if the files have the same size (roughly), we can assume that the files in one chunk would be finished in the same time.

Reply:
I have started making this change but probably won’t finish it until next week.